src/python_testing/TC_OPCREDS_3_5.py

```diff
@@ -0,0 +1,245 @@
+                         "Verify that the DUT returns a CommissioningCompleteResponse with the ErrorCode set to OK"),
+                TestStep(17, "TH1 sends ArmFailSafe command to the DUT with the ExpiryLengthSeconds field set to 0 to verify that the CommissioningComplete command successfully persisted the failsafe context.",
+                         "Verify that DUT sends ArmFailSafeResponse with the ErrorCode set to OK"),
+                TestStep(18, "TH1 reads the NOCs attribute from the Node Operational Credentials cluster using a fabric-filtered read",
+                         "Verify that the returned list has a single entry. Verify that the NOC field matches noc_update2 and the ICAC field matches icac_update2"),
+                TestStep(19, "TH1 reads its fabric index from the CurrentFabricIndex attribute and saves as fabric_idx"),
+                TestStep(20, "TH0 sends the RemoveFabric command with the fabric index set to fabric_idx")
+                ]
+
+    @async_test_body
+    async def test_TC_OPCREDS_3_5(self):
+        opcreds = Clusters.OperationalCredentials
+
+        self.step(1)
+
+        # Start with 10 years for cert validity
+        original_cert_validity = 365*24*60*60*10
+        # We're going to force the use of maximally generated certs because this forces the ICAC to be re-generated each time
+        th1_certificate_authority = self.certificate_authority_manager.NewCertificateAuthority(
+            certificateValidityPeriod=original_cert_validity, maximizeCertChains=True)
+        th1_fabric_admin = th1_certificate_authority.NewFabricAdmin(vendorId=0xFFF1, fabricId=self.matter_test_config.fabric_id + 1)
+        th1_controller_node_id = self.matter_test_config.controller_node_id+1
+        th1 = th1_fabric_admin.NewController(nodeId=th1_controller_node_id)
+
+        success, nocResp, original_cert_chain = await CommissioningBuildingBlocks.AddNOCForNewFabricFromExisting(
+            commissionerDevCtrl=self.default_controller, newFabricDevCtrl=th1,
+            existingNodeId=self.dut_node_id, newNodeId=self.dut_node_id)
+
+        self.step(2)
+        nocs = await self.read_single_attribute_check_success(dev_ctrl=th1, node_id=self.dut_node_id, cluster=opcreds, attribute=opcreds.Attributes.NOCs, fabric_filtered=True)
+        noc_original = nocs[0].noc
+        icac_original = nocs[0].icac
+        asserts.assert_equal(noc_original, original_cert_chain.nocBytes, "Returned NOC does not match generated NOC")
+        asserts.assert_equal(icac_original, original_cert_chain.icacBytes, "Returned ICAC does not match generated ICAC")
+
+        self.step(3)
+        trusted_root_original = await self.read_single_attribute_check_success(
+            dev_ctrl=th1,
+            node_id=self.dut_node_id,
+            cluster=opcreds,
+            attribute=opcreds.Attributes.TrustedRootCertificates)
+        asserts.assert_equal(len(trusted_root_original), 2,
```

Comment:
Sure, do you think it’s better to remove the whole step or just validate that it has x ... entries?

Reply:
```
self.step(3)
rcac_original = original_cert_chain.rcacBytes
trusted_root_original = await self.read_single_attribute_check_success(
            dev_ctrl=th1,
            node_id=self.dut_node_id,
            cluster=opcreds,
            attribute=opcreds.Attributes.TrustedRootCertificates)
# Check if the last element of trusted_root_original matches rcac_original
if trusted_root_original[-1] == rcac_original:
    print("The last element matches rcac_original.")
else:
    print("The last element does not match rcac_original.")
```

I am checking if the last value of trusted_root_original matches rcac_original. There is no assert here—just printing the result. Do you think this is fine, or should I throw in an assert to make it stricter?

Reply:
I noticed that I can use the assert validation by getting the last value and comparing it with the RCAC, so I did it that way. I hope it looks fine to you.

```
self.step(3)
        rcac_original = original_cert_chain.rcacBytes
        trusted_root_certificates = await self.read_single_attribute_check_success(
            dev_ctrl=th1,
            node_id=self.dut_node_id,
            cluster=opcreds,
            attribute=opcreds.Attributes.TrustedRootCertificates)
        if (len(trusted_root_certificates) > 1):
            trusted_root_original = trusted_root_certificates[-1]
        else:
            print("Unexpected error returned list is empty or has no entries")
        asserts.assert_equal(trusted_root_original, rcac_original, "Returned RCAC does not match the TrustedRootCertificates.")
```

Reply:
It was replaced to this, since there is not guaranteed to be always last element.

```
self.step(3)
rcac_original = original_cert_chain.rcacBytes
trusted_root_certificates = await self.read_single_attribute_check_success(
    dev_ctrl=th1,
    node_id=self.dut_node_id,
    cluster=opcreds,
    attribute=opcreds.Attributes.TrustedRootCertificates)
asserts.assert_equal(rcac_original not in trusted_root_certificates, False, "RCAC should match one of the TrustedRootCertificates.")
```

Reply:
This conversation is no longer active since the implementation was completed through the suggestion commit. I will go ahead and close it now.